docs/csharp/nullable-references.md

```diff
@@ -0,0 +1,74 @@
+
+This new feature provides significant benefits over the earlier reference variables where the design intent couldn't be determined from the variable declaration. The compiler didn't provide safety for reference types:
+
+- **A reference could be null**. No warnings are issued when a reference type is initialized to null, or later assigned to null.
+- **A reference is assumed to be not null**. The compiler doesn't issue any warnings when reference types are dereferenced.
+
+A **nullable reference type** is noted using the same syntax as [nullable value types](programming-guide/nullable-types/index.md): a `?` is appended to the type of the variable. For example, the following variable declaration represents a nullable string variable, `name`:
+
+```csharp
+string? name;
+```
+
+Any variable where the `?` is not appended to the type name is a **non-nullable reference type**. That includes all reference type variables in existing code, when you have enabled this feature.
+
+The compiler uses static analysis to determine if a nullable reference is known to be non-null. THe compiler warns you if you dereference a nullable reference when it may be null. You can override this behavior using the **null-forgiving operator** (`!`) following a variable name. For example, if you know the `name` variable isn't null but the compiler issues a warning, you can write the following code to override the compiler's analysis:
+
+```csharp
+name!.Length;
+```
+
+## Nullable contexts
+
+Nullable contexts enable fine-grained control for how the compiler interprets reference type variables. The nullable context of any given source line is `enabled` or `disabled`. You can think of the pre-C# 8 compiler as compiling all your code in a `disabled` nullable context: Any reference type may be null. Warnings are not generated when variables of reference type are de-referenced.
+
+The nullable context is controlled using a new pragma:
+
+- `#nullable enable` sets the nullable context to enabled.
+- `#nullable disable` sets the nullable context to disabled.
+
+The default nullable context is `disabled`. That decision means that your existing code compiles without changes and without generating any new warnings.
+
+The compiler uses the following rules in a disabled nullable context:
+
+- You can't declare nullable references in a disabled context.
+- All reference variables may be assigned to null.
+- No warnings are generated when a variable of a reference type is de-referenced.
+- The null-forgiving operator may not be used in a disabled context.
+
+The behavior is the same as previous versions of C#.
+
+The compiler uses the following rules in an enabled nullable context:
+
+- Any variable of a reference type is a **non-nullable reference type**
+- Any non-nullable reference may be de-referenced safely.
+- Any nullable reference type (noted by `?` after the type in the variable declaration) may be null, and must be verified as not-null before it can be dereferenced.
```

Comment:
you used not null without hyphens before or did you mean to use non-null?

Reply:
Ack. I copied from the ANTLR grammar in the spec. Should be not null.